### DISCUSSION TITLE
Merlin - an Excalibur meta framework

### DISCUSSION THREAD

**Discussion by mattjennings** [excaliburjs/Excalibur#2342] [Show and tell]
Hey all, I wanted to share a little project I've been working on called [Merlin](https://github.com/mattjennings/merlin). It's a "meta framework" that provides an opinionated workflow for creating Excalibur games. (If you've ever used Next.js for React, it's the same idea).

Here are some of the features:

- Uses Vite as the bundler and dev server
- Scenes are automatically loaded and registered according to their filename
   - `src/scenes/level1.ts` automatically gets loaded as a scene called `level1`
- `$res` import alias that automatically parses assets as Excalibur resources and queues them up for loading
   ```js
    import image from '$res/player.png' // -> returns an ImageSource
    
    // no need to add to a loader!
    const sprite = image.toSprite()
   ```

I've converted the [official Sample Platformer](https://github.com/excaliburjs/sample-platformer) project which you can [check out here](https://github.com/mattjennings/merlin/tree/main/examples/sample-platformer). If you want to try it out, run:

```
npx degit mattjennings/merlin/examples/sample-platformer sample-platformer
cd sample-platformer
npm install
npm run dev
``` 

and it _should_ work!

**Please don't use this for production** - I'm still exploring what this framework could look like and things will change/break rapidly (I may not even finish it 😅).

Let me know if you have any thoughts!

**Comment by kamranayub** [comment]
That's awesome, I've always thought it would be cool to do some kind of "router" between scenes that could handle transitions (think Zelda style).

**Reply by mattjennings** [reply]
Agreed! That is definitely something I want to look in to.

**Comment by eonarheim** [comment]
@mattjennings This is so cool! I definitely want to try this out as soon as I can!

This is an excellent way to remove the friction around scenes and loading resources!

**Reply by mattjennings** [reply]
I'll keep this thread updated on any major milestones, but hopefully sooner than later I can have a _decent_ docs page on starting a new project with it. For now though I'm whipping up examples to dogfood it myself, so you can always clone & check those out if you get impatient 😛
